File: webapp/src/lib/subjects.ts
/**
 * NATS subject and KV key builders for LabJack live data, configs and exports.
 *
 * These mirror the builders in `rust-ljm/src/subjects.rs`, which the streamer,
 * archiver and exporter use, so the dashboard subscribes and publishes on the same
 * names. Two live-data layouts exist:
 *
 * - Legacy, keyed by asset number: `<root>.<asset>.data.chNN`, for example
 *   `avenabox.1456.data.ch11`. The root is used exactly as configured.
 * - Structured, keyed by identity: `avenars.<site>.<box>.<source>.live.chNN`, for
 *   example `avenars.i69.i69-mu1.i69-lj2.live.ch11`. Every token is sanitized.
 *
 * The structured layout is used when the root is `avenars` or any of site, box or
 * source ID is set.
 *
 * @module
 */

/** Fields of a LabJack config (KV bucket `avenabox`) that name its subjects. */
export interface LabJackSubjectConfig {
  /** Root subject token, such as `avenabox` (legacy) or `avenars` (structured). */
  nats_subject: string;
  /** Asset number. Used in legacy subjects and as the last source ID fallback. */
  asset_number: number;
  /** LabJack name. Used as the source ID when `source_id` is empty. */
  labjack_name?: string;
  /** Site ID, the `<site>` token of structured subjects. */
  site_id?: string | null;
  /** Edge box ID, the `<box>` token of structured subjects. */
  box_id?: string | null;
  /** Source type. Kept for config compatibility; no builder here reads it. */
  source_type?: string | null;
  /** Source ID, the `<source>` token of structured subjects. */
  source_id?: string | null;
}

/**
 * Converts free-form identity text into a single NATS subject token.
 *
 * Trims and lowercases the text, replaces each run of whitespace, `.` and `/` with
 * one `-`, drops every character other than `a-z`, `0-9`, `_` and `-`, and strips
 * leading and trailing `-`. This keeps `.` and the wildcards `*` and `>` out of the
 * token.
 *
 * @remarks
 * The Rust `sanitize_token` turns each such character into its own `-`, so input
 * with consecutive separators (for example `"a  b"`) gives `a-b` here but `a--b`
 * there.
 *
 * @param raw - Identity text such as a site ID, box ID or LabJack name.
 * @returns The token, or `unknown` if nothing usable remains.
 */
function sanitizeToken(raw: string): string {
  const normalized = raw
    .trim()
    .toLowerCase()
    .replace(/[\s./]+/g, "-")
    .replace(/[^a-z0-9_-]/g, "")
    .replace(/^-+|-+$/g, "");

  return normalized || "unknown";
}

/**
 * Formats an asset number with the three-digit legacy subject width.
 *
 * @param asset - Asset number.
 * @returns The number zero-padded to at least three digits, e.g. `7` gives `007`.
 *   Longer numbers are not truncated.
 */
function padAsset(asset: number): string {
  return String(asset).padStart(3, "0");
}

/**
 * Formats a LabJack channel number as the `chNN` token used in subjects.
 *
 * @param channel - LabJack channel number.
 * @returns `ch` followed by the number zero-padded to at least two digits.
 *
 * @example
 * ```ts
 * padChannel(3);   // "ch03"
 * padChannel(120); // "ch120"
 * ```
 */
export function padChannel(channel: number): string {
  return `ch${String(channel).padStart(2, "0")}`;
}

/**
 * Reports whether a config uses the structured site/box/source layout.
 *
 * @param config - LabJack config.
 * @returns `true` if the trimmed root is `avenars` or any of `site_id`, `box_id` or
 *   `source_id` is non-empty.
 */
function usesStructuredNamespace(config: LabJackSubjectConfig): boolean {
  return config.nats_subject.trim() === "avenars" || Boolean(config.site_id || config.box_id || config.source_id);
}

/**
 * Builds the live-data subject for one LabJack channel.
 *
 * Legacy configs give `<root>.<asset>.data.chNN`, with the root used as configured
 * and the asset padded to three digits. Structured configs give
 * `<root>.<site>.<box>.<source>.live.chNN` with every token sanitized. Missing
 * values fall back to `unknown-site` and `unknown-box`, and the source falls back
 * from `source_id` to `labjack_name` to `asset<NNN>`.
 *
 * @param config - LabJack config.
 * @param channel - LabJack channel number.
 * @returns The subject the streamer publishes that channel's FlatBuffer `Scan`
 *   messages on.
 *
 * @example
 * ```ts
 * liveLabJackChannelSubject({ nats_subject: "avenabox", asset_number: 1456 }, 11);
 * // "avenabox.1456.data.ch11"
 * liveLabJackChannelSubject({
 *   nats_subject: "avenars", asset_number: 1456,
 *   site_id: "i69", box_id: "i69-mu1", source_id: "i69-lj2",
 * }, 11);
 * // "avenars.i69.i69-mu1.i69-lj2.live.ch11"
 * ```
 */
export function liveLabJackChannelSubject(config: LabJackSubjectConfig, channel: number): string {
  if (!usesStructuredNamespace(config)) {
    return `${config.nats_subject}.${padAsset(config.asset_number)}.data.${padChannel(channel)}`;
  }

  const root = sanitizeToken(config.nats_subject);
  const siteId = sanitizeToken(config.site_id || "unknown-site");
  const boxId = sanitizeToken(config.box_id || "unknown-box");
  const sourceId = sanitizeToken(
    config.source_id || config.labjack_name || `asset${padAsset(config.asset_number)}`
  );

  return `${root}.${siteId}.${boxId}.${sourceId}.live.${padChannel(channel)}`;
}

/**
 * Builds a label describing the live subjects of all channels of one source.
 *
 * Same layout and fallbacks as {@link liveLabJackChannelSubject}, with `ch##` in
 * place of the channel token. `ch##` is not a NATS wildcard, so the result is for
 * display only and cannot be subscribed to.
 *
 * @param config - LabJack config.
 * @returns For example `avenars.i69.i69-mu1.i69-lj2.live.ch##`.
 */
export function liveLabJackChannelPattern(config: LabJackSubjectConfig): string {
  if (!usesStructuredNamespace(config)) {
    return `${config.nats_subject}.${padAsset(config.asset_number)}.data.ch##`;
  }

  const root = sanitizeToken(config.nats_subject);
  const siteId = sanitizeToken(config.site_id || "unknown-site");
  const boxId = sanitizeToken(config.box_id || "unknown-box");
  const sourceId = sanitizeToken(
    config.source_id || config.labjack_name || `asset${padAsset(config.asset_number)}`
  );

  return `${root}.${siteId}.${boxId}.${sourceId}.live.ch##`;
}

/**
 * Builds the key of a LabJack config in the KV bucket `avenabox`.
 *
 * Every token is sanitized. Missing values fall back to `unknown-site`,
 * `unknown-box`, and for the source from `source_id` to `labjack_name` to
 * `unknown-source`. Unlike the subject builders there is no `asset<NNN>` fallback.
 *
 * @param config - Identity fields of the config.
 * @returns `<site>.<box>.<source>.config`, e.g. `i69.i69-mu1.i69-lj2.config`.
 */
export function labjackConfigKey(config: {
  site_id?: string | null;
  box_id?: string | null;
  source_id?: string | null;
  labjack_name?: string | null;
}): string {
  const siteId = sanitizeToken(config.site_id || "unknown-site");
  const boxId = sanitizeToken(config.box_id || "unknown-box");
  const sourceId = sanitizeToken(config.source_id || config.labjack_name || "unknown-source");
  return `${siteId}.${boxId}.${sourceId}.config`;
}

/**
 * Builds the export request subject for one archived source.
 *
 * Always uses the structured layout, even for legacy configs, with the same token
 * sanitizing and fallbacks as {@link liveLabJackChannelSubject}. The exporter on the
 * edge box listens on this subject.
 *
 * @param config - LabJack config.
 * @returns `<root>.<site>.<box>.<source>.export.request`, e.g.
 *   `avenars.i69.i69-mu1.i69-lj2.export.request`.
 */
export function archiveExportRequestSubject(config: LabJackSubjectConfig): string {
  const root = sanitizeToken(config.nats_subject);
  const siteId = sanitizeToken(config.site_id || "unknown-site");
  const boxId = sanitizeToken(config.box_id || "unknown-box");
  const sourceId = sanitizeToken(
    config.source_id || config.labjack_name || `asset${padAsset(config.asset_number)}`
  );

  return `${root}.${siteId}.${boxId}.${sourceId}.export.request`;
}

File: rust-ljm/src/subjects.rs
//! NATS subject builders for LabJack live data and archive exports.
//!
//! The streamer publishes each channel's scans on a live subject, the archiver
//! consumes the same subjects from local JetStream and writes Parquet, and the exporter
//! listens on an export request subject. All of them build their subject names here
//! so the layouts stay in one place.
//!
//! Two layouts are supported:
//!
//! * Legacy, keyed by asset number: `avenabox.<asset>.data.chNN`, for example
//!   `avenabox.1456.data.ch11`. The root is used exactly as configured.
//! * Structured, keyed by identity: `avenars.<site>.<box>.<source>.live.chNN`, for
//!   example `avenars.i69.i69-mu1.i69-lj2.live.ch11`. Every token is passed through
//!   [`sanitize_token`].
//!
//! The structured layout is chosen when the root is `avenars` or any of site, box or
//! source ID is set (see [`uses_structured_namespace`]).

#![allow(dead_code)]

/// Converts arbitrary identity text into a single NATS subject token.
///
/// Leading and trailing whitespace is trimmed. ASCII letters and digits are
/// lowercased, `-` and `_` are kept, and whitespace, `.` and `/` become `-`. Every
/// other character is dropped. Leading and trailing `-` are then stripped. This keeps
/// `.` (the subject separator) and the wildcards `*` and `>` out of the token.
///
/// # Arguments
///
/// * `raw` - Identity text such as a site ID, box ID or LabJack name.
///
/// # Returns
///
/// The sanitized token, or `unknown` if nothing usable remains.
///
/// # Examples
///
/// ```text
/// sanitize_token("I69 MU1")   -> "i69-mu1"
/// sanitize_token("site.a/b")  -> "site-a-b"
/// sanitize_token(" *> ")      -> "unknown"
/// ```
fn sanitize_token(raw: &str) -> String {
    let mut out = String::with_capacity(raw.len());
    for ch in raw.trim().chars() {
        if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' {
            out.push(ch.to_ascii_lowercase());
        } else if ch.is_whitespace() || ch == '.' || ch == '/' {
            out.push('-');
        }
    }

    let out = out.trim_matches('-').to_string();
    if out.is_empty() {
        "unknown".to_string()
    } else {
        out
    }
}

/// Formats a LabJack channel number as the `chNN` token used in subjects.
///
/// The number is zero-padded to at least two digits.
///
/// # Arguments
///
/// * `ch` - LabJack channel number.
///
/// # Examples
///
/// ```text
/// pad_channel(3)   -> "ch03"
/// pad_channel(11)  -> "ch11"
/// pad_channel(120) -> "ch120"
/// ```
pub fn pad_channel(ch: u8) -> String {
    format!("ch{ch:02}")
}

/// Formats an asset number with the three-digit legacy subject width.
///
/// The number is zero-padded to at least three digits; longer numbers are not
/// truncated.
///
/// # Arguments
///
/// * `n` - Asset number.
///
/// # Examples
///
/// ```text
/// pad_asset(7)    -> "007"
/// pad_asset(1456) -> "1456"
/// ```
pub fn pad_asset(n: u32) -> String {
    format!("{n:03}")
}

/// Decides whether a config should use the structured subject namespace.
///
/// # Arguments
///
/// * `nats_subject` - Configured subject root, for example `avenabox` or `avenars`.
/// * `site_id` - Configured site ID, if any.
/// * `box_id` - Configured box ID, if any.
/// * `source_id` - Configured source ID, if any.
///
/// # Returns
///
/// `true` if the trimmed root is exactly `avenars` or any of the three IDs is `Some`
/// (even an empty string). A LabJack name alone does not switch to the structured
/// layout.
fn uses_structured_namespace(
    nats_subject: &str,
    site_id: Option<&str>,
    box_id: Option<&str>,
    source_id: Option<&str>,
) -> bool {
    nats_subject.trim() == "avenars" || site_id.is_some() || box_id.is_some() || source_id.is_some()
}

/// Builds the live-data subject for one LabJack channel.
///
/// Legacy configs use `<root>.<asset>.data.<channel>`, with the root used as given.
/// Structured configs use `<root>.<site>.<box>.<source>.live.<channel>`, with every
/// token passed through [`sanitize_token`]. In the structured layout a missing site or
/// box becomes `unknown-site` or `unknown-box`, and the source is the first of
/// `source_id`, `labjack_name` or `asset<NNN>` that is set.
///
/// # Arguments
///
/// * `nats_subject` - Subject root, for example `avenabox` or `avenars`.
/// * `asset` - Asset number. Used in the legacy layout and as the last-resort source
///   token in the structured layout.
/// * `channel` - LabJack channel number, formatted by [`pad_channel`].
/// * `site_id` - Site ID for the structured layout.
/// * `box_id` - Box ID for the structured layout.
/// * `labjack_name` - LabJack name, used as the source when `source_id` is `None`.
/// * `source_type` - Accepted for config compatibility; not used in the subject.
/// * `source_id` - Source ID for the structured layout.
///
/// # Examples
///
/// ```text
/// ("avenabox", 1456, 11, None, None, None, None, None)
///     -> "avenabox.1456.data.ch11"
/// ("avenars", 1456, 11, Some("i69"), Some("i69-mu1"), Some("i69-lj2"), None, None)
///     -> "avenars.i69.i69-mu1.i69-lj2.live.ch11"
/// ("avenars", 42, 3, None, None, None, None, None)
///     -> "avenars.unknown-site.unknown-box.asset042.live.ch03"
/// ```
pub fn live_labjack_channel_subject(
    nats_subject: &str,
    asset: u32,
    channel: u8,
    site_id: Option<&str>,
    box_id: Option<&str>,
    labjack_name: Option<&str>,
    source_type: Option<&str>,
    source_id: Option<&str>,
) -> String {
    if !uses_structured_namespace(nats_subject, site_id, box_id, source_id) {
        return format!(
            "{}.{}.data.{}",
            nats_subject,
            pad_asset(asset),
            pad_channel(channel)
        );
    }

    let root = sanitize_token(nats_subject);
    let site_id = sanitize_token(site_id.unwrap_or("unknown-site"));
    let box_id = sanitize_token(box_id.unwrap_or("unknown-box"));
    let source = source_id
        .or(labjack_name)
        .map(str::to_string)
        .unwrap_or_else(|| format!("asset{}", pad_asset(asset)));
    let source_id = sanitize_token(&source);

    let _ = source_type;

    format!(
        "{root}.{site_id}.{box_id}.{source_id}.live.{}",
        pad_channel(channel)
    )
}

/// Builds the JetStream subject wildcard for all live channels from one source.
///
/// Legacy configs get `<root>.*.data.*`, which matches every asset under the root.
/// Structured configs get `<root>.<site>.<box>.<source>.live.*`, sanitized the same way
/// as [`live_labjack_channel_subject`].
///
/// The structured source fallback differs from [`live_labjack_channel_subject`]: when
/// neither `source_id` nor `labjack_name` is set this returns `unknown-source`, while
/// the channel subject falls back to `asset<NNN>`. Such a wildcard does not match the
/// channel subjects built from the same config.
///
/// # Arguments
///
/// * `nats_subject` - Subject root, for example `avenabox` or `avenars`.
/// * `site_id` - Site ID for the structured layout.
/// * `box_id` - Box ID for the structured layout.
/// * `labjack_name` - LabJack name, used as the source when `source_id` is `None`.
/// * `source_type` - Accepted for config compatibility; not used in the subject.
/// * `source_id` - Source ID for the structured layout.
///
/// # Examples
///
/// ```text
/// ("avenabox", None, None, None, None, None)
///     -> "avenabox.*.data.*"
/// ("avenars", Some("i69"), Some("i69-mu1"), Some("i69-lj2"), None, None)
///     -> "avenars.i69.i69-mu1.i69-lj2.live.*"
/// ```
pub fn live_labjack_stream_subject(
    nats_subject: &str,
    site_id: Option<&str>,
    box_id: Option<&str>,
    labjack_name: Option<&str>,
    source_type: Option<&str>,
    source_id: Option<&str>,
) -> String {
    if !uses_structured_namespace(nats_subject, site_id, box_id, source_id) {
        return format!("{nats_subject}.*.data.*");
    }

    let root = sanitize_token(nats_subject);
    let site_id = sanitize_token(site_id.unwrap_or("unknown-site"));
    let box_id = sanitize_token(box_id.unwrap_or("unknown-box"));
    let source_id = sanitize_token(source_id.or(labjack_name).unwrap_or("unknown-source"));

    let _ = source_type;

    format!("{root}.{site_id}.{box_id}.{source_id}.live.*")
}

/// Builds the NATS request subject that the exporter serves archive exports on.
///
/// Always uses the structured layout `<root>.<site>.<box>.<source>.export.request`,
/// whatever the root, with every token passed through [`sanitize_token`]. Missing IDs
/// become `unknown-site`, `unknown-box` and `unknown-source`. Unlike the live subjects,
/// there is no fallback to a LabJack name or asset number.
///
/// # Arguments
///
/// * `nats_subject` - Subject root, for example `avenars`.
/// * `site_id` - Site ID.
/// * `box_id` - Box ID.
/// * `source_type` - Accepted for config compatibility; not used in the subject.
/// * `source_id` - Source ID.
///
/// # Examples
///
/// ```text
/// ("avenars", Some("i69"), Some("i69-mu1"), Some("labjack"), Some("i69-lj2"))
///     -> "avenars.i69.i69-mu1.i69-lj2.export.request"
/// ```
pub fn archive_export_request_subject(
    nats_subject: &str,
    site_id: Option<&str>,
    box_id: Option<&str>,
    source_type: Option<&str>,
    source_id: Option<&str>,
) -> String {
    let root = sanitize_token(nats_subject);
    let site = sanitize_token(site_id.unwrap_or("unknown-site"));
    let box_id = sanitize_token(box_id.unwrap_or("unknown-box"));
    let source_id = sanitize_token(source_id.unwrap_or("unknown-source"));

    let _ = source_type;

    format!("{root}.{site}.{box_id}.{source_id}.export.request")
}

/// Checks whether an existing stream subject is compatible with a desired one.
///
/// An exact match is compatible. When the desired subject is a legacy wildcard
/// `<prefix>.*.data.*`, an existing subject is also compatible if it starts with
/// `<prefix>.` and ends with `.data.*`, so an older per-asset pattern such as
/// `avenabox.1456.data.*` is accepted. Anything else is incompatible.
///
/// # Arguments
///
/// * `existing` - A subject already configured on the JetStream stream.
/// * `desired_namespace` - The subject the current config wants, typically from
///   [`live_labjack_stream_subject`].
///
/// # Examples
///
/// ```text
/// ("avenabox.1456.data.*", "avenabox.*.data.*")         -> true
/// ("avenabox.*.data.*", "avenabox.*.data.*")            -> true
/// ("avenars.i69.a.b.live.*", "avenars.i69.a.c.live.*")  -> false
/// ```
pub fn stream_subject_is_compatible(existing: &str, desired_namespace: &str) -> bool {
    if existing == desired_namespace {
        return true;
    }

    if let Some(prefix) = desired_namespace.strip_suffix(".*.data.*") {
        return existing.starts_with(&format!("{prefix}.")) && existing.ends_with(".data.*");
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Legacy configs keep the `<root>.<asset>.data.chNN` layout.
    #[test]
    fn legacy_subjects_are_preserved() {
        assert_eq!(
            live_labjack_channel_subject("avenabox", 1456, 11, None, None, None, None, None),
            "avenabox.1456.data.ch11"
        );
        assert_eq!(
            live_labjack_stream_subject("avenabox", None, None, None, None, None),
            "avenabox.*.data.*"
        );
    }

    /// Structured configs produce site, box and source tokens in every subject kind.
    #[test]
    fn structured_subjects_include_site_box_source_and_channel() {
        assert_eq!(
            live_labjack_channel_subject(
                "avenars",
                1456,
                11,
                Some("i69"),
                Some("i69-mu1"),
                Some("i69-lj2"),
                None,
                None,
            ),
            "avenars.i69.i69-mu1.i69-lj2.live.ch11"
        );
        assert_eq!(
            live_labjack_stream_subject(
                "avenars",
                Some("i69"),
                Some("i69-mu1"),
                Some("i69-lj2"),
                None,
                None,
            ),
            "avenars.i69.i69-mu1.i69-lj2.live.*"
        );
        assert_eq!(
            archive_export_request_subject(
                "avenars",
                Some("i69"),
                Some("i69-mu1"),
                Some("labjack"),
                Some("i69-lj2"),
            ),
            "avenars.i69.i69-mu1.i69-lj2.export.request"
        );
    }
}
